tests/e2e/improved_csi_idempotency.go

```diff
@@ -0,0 +1,519 @@
+		2. Create a PVCs using SC
+		3. Restart CSI driver when VC task has started
+		4. Wait for PVCs to be bound
+		5. Delete PVCs and SC
+		6. Verify no orphan volumes are left
+	*/
+	ginkgo.It("create volume when CSI restarts - idempotency", func() {
+		createVolumeWithServiceDown("CSI", namespace, client, storagePolicyName,
+			scParameters, volumeOpsScale)
+	})
+
+	/*
+		Extend volume when csi restarts
+		1. Create a SC using a thick provisioned policy
+		2. Create a PVCs using SC and wait for PVCs to be bound
+		3. Extend pvcs size
+		4. Restart CSI when VC task is ongoing
+		5. Verify resize was successful
+		6. Delete pvcs and SC
+		7. Verify no orphan volumes are left
+	*/
+	ginkgo.It("extend volume when csi restarts - idempotency", func() {
+		// Check nfs4FSType for file volumes
+		kind := os.Getenv("ACCESS_MODE")
+		if strings.TrimSpace(string(kind)) == "RWX" {
+			ginkgo.Skip("File Volume extend is not supported, skipping the test")
+		}
+		extendVolumeWithServiceDown("CSI", namespace, client, storagePolicyName, scParameters,
+			volumeOpsScale, true)
+	})
+
+	/*
+		Extend volume when CNS goes down
+		1. Create a SC using a thick provisioned policy
+		2. Create a PVCs using SC and wait for PVC to be bound
+		3. Extend pvcs and wait for vc task to start
+		4. Bring down vsan-health service for 6mins
+		5. Restart vsan-health service
+		6. Verify resize was successful
+		7. Delete pvcs and SC
+		8. Verify no orphan volumes are left
+	*/
+	ginkgo.It("extend volume when CNS goes down - idempotency", func() {
+		// Check if it is file volumes setups
+		kind := os.Getenv("ACCESS_MODE")
+		if strings.TrimSpace(string(kind)) == "RWX" {
+			ginkgo.Skip("File Volume extend is not supported, skipping the test")
+		}
+		extendVolumeWithServiceDown(vsanhealthServiceName, namespace, client, storagePolicyName, scParameters,
+			volumeOpsScale, true)
+	})
+})
+
+// createVolumeWithServiceDown creates the volumes and immediatly stops the services and wait for
+// the service to be up again and validates the volumes are bound
+func createVolumeWithServiceDown(serviceName string, namespace string, client clientset.Interface,
+	storagePolicyName string, scParameters map[string]string, volumeOpsScale int) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	ginkgo.By(fmt.Sprintf("Invoking Test for create volume when %v goes down", serviceName))
+	var storageclass *storagev1.StorageClass
+	var persistentvolumes []*v1.PersistentVolume
+	var pvclaims []*v1.PersistentVolumeClaim
+	var err error
+	var isServiceStopped bool
+	var fullSyncWaitTime int
+	pvclaims = make([]*v1.PersistentVolumeClaim, volumeOpsScale)
+
+	// Decide which test setup is available to run
+	if vanillaCluster {
+		ginkgo.By("CNS_TEST: Running for vanilla k8s setup")
+		scParameters[scParamStoragePolicyName] = "Management Storage Policy - Regular"
+		// Check if it is file volumes setups
+		kind := os.Getenv("ACCESS_MODE")
```

Comment:
can you do it once for the whole run, just the way we do it for `vanillaCluster`

Reply:
I don't think it is done, can you recheck